README.md
# Site.js
## Small web construction set.

[![Person lying on the ground, working on a laptop with the Site.js logo on screen](images/person.svg)](https://sitejs.org)

## Develop, test, sync, and deploy (using a single tool that comes in a single binary).

__Site.js is a [small](https://small-tech.org/about#small-technology) personal web tool for Linux, macOS, and Windows 10.__

Most tools today are built for startups and enterprises. Site.js is built for people.

## Like this? Fund us!

[Small Technology Foundation](https://small-tech.org) is a tiny, independent not-for-profit.

We exist in part thanks to patronage by people like you. If you share [our vision](https://small-tech.org/about/#small-technology) and want to support our work, please [become a patron or donate to us](https://small-tech.org/fund-us) today and help us continue to exist.

## Feature Highlights

  - __Just works.__ No configuration; [get started in seconds](https://sitejs.org/#get-started).

  - __Free as in freedom.__ And small as in [small tech](https://small-tech.org/about/#small-technology).

  - __Seamless single binary [install](#install)__ (thanks to [Nexe](https://github.com/nexe/nexe)).

  - __Secure by default.__

    __At localhost:__ Automatically provisions locally-trusted TLS for development (courtesy of [mkcert](https://github.com/FiloSottile/mkcert) seamlessly integrated via [Auto Encrypt Localhost](https://github.com/small-tech/auto-encrypt-localhost)).

    __And everywhere else:__ Automatically provisions globally-trusted TLS for staging and production (courtesy of [Let’s Encrypt](https://letsencrypt.org/) seamlessly integrated via [Auto Encrypt](https://github.com/small-tech/auto-encrypt) and [systemd](https://freedesktop.org/wiki/Software/systemd/)).

    Your server will score [an A+](https://www.ssllabs.com/ssltest/analyze.html?d=sitejs.org) on the [SSL Labs SSL Server Test](https://www.ssllabs.com/ssltest).

  - __Supports the creation of static web sites, dynamic web sites, and hybrid sites__ (via integrated [Node.js](https://nodejs.org/) and [Express](https://expressjs.com)).

  - __Includes a fast and simple database__ ([JSDB](https://github.com/small-tech/jsdb)).

  - __Supports [wildcard routes](#wildcard-routes)__ for purely client-side specialisation using path-based parameters.

  - __Supports [DotJS](#dotjs) for dynamic routes.__ (DotJS is PHP-like simple routing for Node.js introduced by Site.js for quickly prototyping and building dynamic sites).

  - __Includes [Hugo static site generator](#static-site-generation).__

  - __Includes [one command installation of production-ready Owncast server](#owncast-integration).__

  - __[Sync](#sync) to deploy__ (uses rsync for quick deployments). Can also [Live Sync](#live-sync) for live blogging, etc. For sites that implement the [Small Web](https://ar.al/2020/08/07/what-is-the-small-web/) conventions, you can also use the simplified [pull and push commands](#pull-and-push).

  - __Has privacy-respecting [ephemeral statics](#ephemeral-statistics)__. Gives you insight into how your site is being used, not into the people using it.

  - __Supports [WebSockets](#websocket-wss-routes)__ (via integrated [express-ws](https://github.com/HenningM/express-ws), which itself wraps [ws](https://github.com/websockets/ws)).

  - __Can be used as a proxy server__ (via integrated [http-proxy-middleware](https://github.com/chimurai/http-proxy-middleware)).

  - __Supports [an evergreen web](#native-support-for-an-evergreen-web).__

    [The archival cascade](#the-archival-cascade) and [Native 404 → 302 support](#native-404--302-support) help you migrate and evolve your existing sites using Site.js without breaking existing links.

  - __Live reload__ on static pages.

  - __Automatic server reload__ when the source code of your dynamic routes change.

  - __Auto updates__ of production servers.

  <ins>Note:</ins> Production use via startup daemon is only supported on Linux distributions with systemd.

## Install

Copy and paste the following commands into your terminal:

__(Note: all commands should be run in your regular account, not as root.)__ (As of 15.4.0, Site.js will refuse to run if launched from the root account.)

### Native binaries

__Before you pipe any script into your computer, always view the source code ([Linux and macOS](https://should-i-pipe.it/https://sitejs.org/install), [Windows](https://should-i-pipe.it/https://sitejs.org/install.txt)) and make sure you understand what it does.__

#### Linux

```shell
wget -qO- https://sitejs.org/install | bash
```

(To use curl instead, see the macOS instructions, below.)

#### macOS

```shell
curl -s https://sitejs.org/install | bash
```

#### Windows 10 with PowerShell running under Windows Terminal

```shell
iex(iwr -UseBasicParsing https://sitejs.org/install.txt).Content
```

### Node.js

```shell
npm i -g @small-tech/site.js
```

### Alpha and Beta channels

On Linux and macOS, in addition to the release build channel, there is also an alpha build and beta build channel available. Pass either `alpha` or `beta` as an argument to the Bash pipe to install the latest build from the respective channel.

For example, to install the latest beta build on Linux:

```shell
wget -qO- https://sitejs.org/install | bash -s -- beta
```

__Note:__ On Macs, `wget` is not installed by default but `curl` is so you can use that instead:

```shell
curl -s https://sitejs.org/install | bash -s -- beta
```

Alpha builds are strictly for local testing and should not, under any circumstances, be used in production. We do not test Alpha builds in production.

Servers deployed using release builds check for updates every six hours whereas beta builds check every 10 minutes.

Note that the latest alpha or beta build available may be older than the latest release build. You can check the date on the build via the `version` command.

## System Requirements

### Linux

Any recent Linux distribution should work. However, Site.js is most thoroughly tested on Ubuntu-esque distributions (development and staging) and Ubuntu LTS (production).

There are builds available for x64, ARM, and ARM64.

For production use, [systemd](https://en.wikipedia.org/wiki/Systemd) is required.

### macOS

macOS 10.14.x Mojave and macOS 10.15.x Catalina are supported (the latter as of Site.js 12.5.1).

_Production use is not possible under macOS._

### Windows 10

The current version of Windows 10 is supported with PowerShell running under [Windows Terminal](https://github.com/Microsoft/Terminal).

__Windows Subsystem for Linux (WSL) is _not_ supported.__ (You can install and run Site.js under WSL but seamless TLS certificate handling for local servers will not work out of the box as WSL and Windows 10 do not share certificate stores. If you do want to use Site.js under WSL, you have to first install Site.js on Windows 10 and run a local server (`site`) to create the certificate authority and certificates, then install and run Site.js under WSL and then manually copy the contents of `~/.small-tech.org/site.js/tls/local/` from Windows 10 to WSL.)

_Production use is not possible under Windows._

## Dependencies

Site.js tries to seamlessly install the dependencies it needs when run. That said, there are certain basic components it expects on a Linux-like system. These are:

  - `sudo`
  - `bash` (on Linux, macOS, etc.) or `PowerShell` running under [Windows Terminal](https://github.com/Microsoft/Terminal) (on Windows 10).
  - `wget` or `curl` (on Linux and macOS) are required to download the installation script when installing Site.js using the one-line installation command. On Linux, you can install either via your distribution’s package manager (e.g., `sudo apt install wget` on Ubuntu-like systems). macOS comes with `curl` installed.

If it turns out that any of these prerequisites are a widespread cause of first-run woe, we can look into having them installed automatically in the future. Please [open an issue](https://github.com/small-tech/site.js/issues) if any of these affects you during your deployments or in everyday use.

### Automatically-installed dependencies

__For production use, passwordless sudo is required.__ On systems where the sudo configuration directory is set to `/etc/sudoers.d`, Site.js will automatically install this rule. On other systems, you might have to [set it up yourself](https://serverfault.com/questions/160581/how-to-setup-passwordless-sudo-on-linux).

__For localhost servers__, the bundled [mkcert](https://github.com/FiloSottile/mkcert) requires [certutil](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/tools/NSS_Tools_certutil) and the [Network Security Services](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS) (NSS) dynamic libraries. Site.js will attempt to automatically install the required libraries using popular package managers. <strike>Please note that this will fail on PinePhones running UBPorts as NSS is missing from the apt package manager for that distribution.</strike> ([The PinePhone issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1652739) has been resolved.)

## Update (as of version 12.9.5; properly functioning as of version 12.9.6)

To seamlessly update the native binary if a newer version exists:

```shell
site update
```

This command will automatically restart a running Site.js daemon if one exists. If you are running Site.js as a regular process, it will continue to run and you will run the newer version the next time you launch a regular Site.js process.

__Note:__ There is a bug in the semantic version comparison in the original release with the update feature (version 12.9.5) that will prevent upgrades between minor versions (i.e., between 12.9.5 and 12.10.x and beyond). This was fixed in version 12.9.6. If you’re still on 12.9.5 and you’re reading this after we’ve moved to 12.10.0 and beyond, please stop Site.js if it’s running and [install the latest Site.js](#install) manually.

## Automatic updates in production (as of version 12.10.0)

[Production servers](#production) started with the `enable` command will automatically check for updates on first launch and then again at a set interval (currently every 6 hours) and update themselves as and when necessary.

This is a primary security feature given that Site.js is meant for use by individuals, not startups or enterprises with operations teams that can (in theory, at least) maintain servers with the latest updates.

## Uninstall

To uninstall the native binary (and any created artifacts, like TLS certificates, systemd services, etc.):

```shell
site uninstall
```

## Use

### Development (servers @localhost)

#### Regular server

Start serving the current directory at https://localhost as a regular process using locally-trusted certificates:

```shell
$ site
```

Note that if your current working directory is inside a special subfolder of your site (`.dynamic`, `.hugo`, `.wildcard`, `.db`) Site.js (as of version 15.4.0) magically does the right thing and serves the site root instead of the folder you’re in. If you really do want to serve one of these folders or a subfolder thereof, specifically state your intent by passing the current folder (`.`) as an argument.

The above caveat aside, the command above is a shorthand for the full form of the `serve` command:

```shell
$ site serve . @localhost:443
```

__Note:__ As of 15.4.0, Site.js will refuse to serve the root directory or your home directory for security reasons.

#### To serve on a different port

Just specify the port explicitly as in the following example:

```shell
$ site @localhost:666
```

That, again, is shorthand for the full version of the command, which is:

```shell
$ site serve . @localhost:666
```

#### A note about running in containers

If your development machine runs an ‘immutable’ OS like [Fedora Silverblue](https://silverblue.fedoraproject.org/), you will likely be running Site.js inside a container (e.g., in a [podman](https://podman.io/) container via [DistroBox](https://distrobox.privatedns.org).)

If so, make sure you [run the container with real root](https://distrobox.privatedns.org/useful_tips.html#run-the-container-with-real-root) and make sure that your container is [set up for using an init system inside it](https://distrobox.privatedns.org/useful_tips.html#using-init-system-inside-a-distrobox) (for systemd-related functionality).

For Distrobox, for example, this means passing the `--root` and `--init` flags to the `distrobox create` command. e.g., The following will create a rootful Ubuntu container using the latest Ubuntu image and call it _ubuntu-root-init_:

```shell
distrobox create --image ubuntu:latest --name ubuntu-root-init --root --init
```

All [tests](#build-and-test-from-source) should pass when run within such a container that is also accessible from your container’s hostname.

#### Accessing your local server over the local area network

You can access local servers via their IPv4 address over a local area network.

This is useful when you want to test your site with different devices without having to expose your server over the Internet using a service like ngrok. For example, if your machine’s IPv4 address on the local area network is 192.168.2.42, you can just enter that IP to access it from, say, your iPhone.

To access your local machine from a different device on your local area network, you must transfer the public key of your generated local root certificate authority to that device and install and trust it.

For example, if you’re on an iPhone, hit the `/.ca` route in your browser:

```
http://192.168.2.42/.ca
```

The browser will download the local root certificate authority’s public key and prompt you to install the profile on your iPhone. You then have to go to Settings → Profile Downloaded → Tap Install when the Install Profile pop-up appears showing you the mkcert certificate you downloaded. Then, go to Settings → General → About → Certificate Trust Settings → Turn on the switch next to the mkcert certificate you downloaded. You should now be able to hit `https://192.168.2.42` and see your site from your iPhone.

You can also tranfer your key to your other devices manually. You can find the key at `~/.small-tech/site.js/tls/local/rootCA.pem` after you’ve created a local server at least once. For more details on transferring your key to other devices, please refer to [the relevant section in the mkcert documentation](https://github.com/FiloSottile/mkcert#mobile-devices).

#### Proxy server

You can use Site.js as a reverse proxy for HTTP and WebSocket connections. This is useful if you have a web app written in any language that only supports HTTP (not TLS) that you want to deploy securely.

For example, the following is a simple HTTP server written in Python 3 (_server.py_) that runs insecurely on port 3000:

```python
from http.server import HTTPServer, BaseHTTPRequestHandler

class MyRequestHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        self.send_response(200)
        self.end_headers()
        self.wfile.write(b'Hello, from Python!')

server = HTTPServer(('localhost', 3000), MyRequestHandler)
server.serve_forever()
```

Run it (at http://localhost:3000) with:

```shell
$ python3 server
```

Then, proxy it securely from https://localhost using:

```shell
$ site :3000
```

Again, this is a convenient shortcut. The full form of this command is:

```shell
$ site serve :3000 @localhost:443
```

This will create and serve the following proxies:

  * http://localhost:3000 → https://localhost
  * ws://localhost:3000 → wss://localhost

### Testing (servers @hostname)

#### Regular server

Start serving the _my-site_ directory at your _hostname_ as a regular process using globally-trusted Let’s Encrypt certificates:

```shell
$ site my-site @hostname
```

Note that as of 13.0.0, Site.js will refuse to start the server if your hostname (or the domain you specified manually using the `--domain` option and any aliases you may have specified using the `--aliases` option) fails to resolve or is unreachable. This should help you diagnose and fix typos in domain names as well as DNS misconfiguration and propagation issues. As of 14.1.0, you can use the `--skip-domain-reachability-check` flag to override this behaviour and skip the pre-flight checks.

#### Proxy server

Start serving `http://localhost:1313` and `ws://localhost:1313` at your _hostname_:

```shell
$ site :1313 @hostname
```

#### macOS notes

To set your hostname under macOS (e.g., to `example.small-tech.org`), run the following command:

```shell
$ sudo scutil --set HostName example.small-tech.org
```

#### Windows 10 notes

On Windows 10, you must add quotation marks around `@hostname` and `@localhost`. So the first example, above, would be written in the following way on Windows 10:

```shell
$ site my-site "@hostname"
```

Also, Windows 10, unlike Linux and macOS, does not have the concept of a hostname. The closest thing to it is your _full computer name_. Setting your full computer name is a somewhat convoluted process so we’ve documented it here for you.

##### How to set your full computer name on Windows 10

Say you want to set your hostname to `my-windows-laptop.small-tech.org`:

1. Control Panel → System And Security → System → Change Settings link (next to Computer name) → [Change…] Button
2. Under Computer name, enter your _subdomain_ (`my-windows-laptop`)
3. [More…] Button → enter your _domain name_ (`small-tech.org`) in the Primary DNS suffix of this computer field.
4. Press the various [OK] buttons to dismiss the various modal dialogues and restart your computer.

#### Making your server public

Use a service like [ngrok](https://ngrok.com/) (Pro+) to point a custom domain name to your temporary staging server. Make sure you set your `hostname` file (e.g., in `/etc/hostname` or via `hostnamectl set-hostname <hostname>` or the equivalent for your platform) to match your domain name. The first time you hit your server via your hostname it will take a little longer to load as your Let’s Encrypt certificates are being automatically provisioned by Auto Encrypt.

When you start your server, it will run as a regular process. It will not be restarted if it crashes or if you exit the foreground process or restart the computer.

### Deployment

#### Pull and push

As of version 14.4.0, you can use the simplified `pull` and `push` commands if your local and remote setup adheres to the following Small Web conventions:

##### Local

  - The name of your local working folder is the same as your domain (if not, specify the domain using the `--domain` oiption)
  - Your SSH key is either found at `~/.ssh/id_{your domain}_ed25519` or you have an _id_25519_ or _id_rsa_ file in your `~/.ssh` folder. (The former is a Small Web convention, the latter is a fallback general convention.)

##### Remote

  - __Account name:__ `site`
  - __Folder being served:__ `/home/site/public`

If those requirements are met, from within your site’s folder on your local machine, you can pull (download) your site using:

```shell
site pull
```

And you can push (deploy) your site using:

```shell
site push
```

The legacy `sync` command will continue to work as before and is documented below.

#### Sync

Site.js can help you deploy your site to your live server with its sync feature.

```shell
$ site my-demo --sync-to=my-demo.site
```

The above command will:

  1. Generate any Hugo content that might need to be generated.
  2. Sync your site from the local _my-demo_ folder via rsync over ssh to the host _my-demo.site_.

Without any customisations, the sync feature assumes that your account on your remote server has the same name as your account on your local machine and that the folder you are watching (_my-demo_, in the example above) is located at _/home/your-account/my-demo_ on the remote server. Also, by default, the contents of the folder will be synced, not the folder itself. You can change these defaults by specifying a fully-qualified remote connection string as the `--sync-to` value.

The remote connection string has the format:

```
remoteAccount@host:/absolute/path/to/remoteFolder
```

For example:

```shell
$ site my-folder --sync-to=someOtherAccount@my-demo.site:/var/www
```

If you want to sync not the folder’s contents but the folder itself, use the `--sync-folder-and-contents` flag. e.g.,

```shell
$ site my-local-folder --sync-to=me@my.site:my-remote-folder --sync-folder-and-contents
```

The above command will result in the following directory structure on the remote server: _/home/me/my-remote-folder/my-local-folder_. It also demonstrates that if you specify a relative folder, Site.js assumes you mean the folder exists in the home directory of the account on the remote server.

(As of 15.4.0) If the sync command cannot connect in 5 seconds, it will time out. If this happens, check that you have the correct host and account details specified. If you do, there might be a problem with your connection.

(As of 16.1.0) It’s a common mistake to start the sync without specifying the `--sync-from` option when not in the root of your site but in one of the well-known subfolders (e.g., `.hugo` if you’re working on a Hugo site or your `.dynamic` folder if you happen to be in it because you’re working on a site that uses DotJS.) In these instances, Site.js will detect the mistake and understand that you want to sync the site, not the subfolder and behave accordingly. If you _really_ want to sync one of the well-known subfolders for some reason, then specifically specify it by setting `--sync-from=.`. Note that this magic rewriting of the sync path doesn’t happen any time you specify a folder explicitly using the `--sync-from` option.

#### Live Sync

With the Live Sync feature, you can have Site.js watch for changes to your content and sync them to your server in real-time (e.g., if you want to live blog something or want to keep a page updated with local data you’re collecting from a sensor).

To start a live sync server, provide the `--live-sync` flag to your sync request.

For example:

```shell
$ site my-demo --sync-to=my-demo.site --live-sync
```

The above command will start a local development server at _https://localhost_. Additionally, it will watch the folder _my-demo_ for changes and sync any changes to its contents via rsync over ssh to the host _my-demo.site_.


### Production

__Available on Linux distributions with systemd (most Linux distributions, but [not these ones](https://sysdfree.wordpress.com/2019/03/09/135/) or on macOS or Windows).__

__For production use, passwordless sudo is required.__ On systems where the sudo configuration directory is set to `/etc/sudoers.d`, Site.js will automatically install this rule. On other systems, you might have to [set it up yourself](https://serverfault.com/questions/160581/how-to-setup-passwordless-sudo-on-linux).

__Please make sure that you are NOT running as root.__ (As of 15.4.0, Site.js will refuse to run if launched from the root account.)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [17.7.1] - 2022-06-14

### Fixed

  - The HTTP Proxy feature no longer has the change origin setting set to true. This fixes federation failing with `crypto/rsa: verification error` errors (e.g., on remote follow attempts) when proxying Owncast.

## [17.7.0] - 2022-06-11

### Updated

  - Owncast to version 0.0.11. 
  
    This version brings fediverse functionality to Owncast servers that you can set up by running `site enable --owncast`.

## [17.6.0] - 2022-06-10

### Updated

  - [@small-tech/https](https://github.com/small-tech/https) to [version 2.2.0](https://github.com/small-tech/https/blob/2.x/CHANGELOG.md#220---2022-06-07). 
  
    Includes Auto Encrypt version to 2.3.0. This updates the certificate signing request (CSR) signature algorithm from the obsolete SHA-1 to SHA-256. (Let’s Encrypt will beging to reject certificate requests signed with SHA-1 on September 15, 2022. See https://community.letsencrypt.org/t/rejecting-sha-1-csrs-and-validation-using-tls-1-0-1-1-urls/175144)

## [17.5.0] - 2022-01-31

### Updated

  - Owncast to version 0.0.10.

## [17.4.1] - 2022-01-05

### Improved

  - Added links to the Site.js site and documentation in the statistics view with further information on the server and how ephemeral statistics work.

## [17.4.0] - 2022-01-05

### Improved

  - Statistics view now displays referrer URLs not IP addresses (so you can see where people link to your site from). Note that these are based what browsers tell us so they can be spoofed.

## [17.3.6] - 2022-01-05

### Fixed

  - Statistics: fix regular expression so all 32-digit hexadecimal strings are replaced in a string even if there are multiple.

## [17.3.5] - 2022-01-05

### Improved

  - Statistics: any 32-digit hexadecimal string (e.g., as used in the admin pattern or for ids) found in Urls is masked so as not to leak them in public statistics.

  - The statistics route itself is masked in the browser (requires JavaScript to be on) so it is not inadvertently revealed in screenshots or browser history (this does not mean you should access private routes on public computers as browsers may still cache URLs).

  - Updated readme with JavaScript snippet you can use on your own sites to mask routes with random cryptographically-secure path fragments to protect them from being inadvertently leaked in screenshots.

  (Please see the readme for the full security model of this unconventional method if you’re going to make use of it on your own sites.)

## [17.3.4] - 2022-01-01

### Fixed

  - Upgrade JSDB to 1.2.2; fixes crash during copy of complex data types (e.g., Date) from one table to another.

## [17.3.3] - 2021-12-31

### Fixed

  - __Security issue__: due to a misconfiguration of how [dotfiles](https://github.com/pillarjs/send#dotfiles) are handled in the [send](https://www.npmjs.com/package/send) package that’s used within the static file serving functionality, if an attacker knew the exact path to a file in a dotfile directory (for example, the default `.db` [JSDB](https://github.com/small-tech/jsdb#javascript-database-jsdb)), they could download the files (e.g., the database tables) within it. This was due to the default dotfiles setting in send being “similar to `'ignore'`”, not exactly like it. The fix was to explicitly set it to `'ignore'` so that files within dotfile directories are not served.

## [17.3.2] - 2021-12-29

### Fixed

  - Upgrade JSDB to 1.2.1; fixes database corruption when serialising certain edge-case object keys (like strings that start with a number or contain all digits and have left padding, etc.)

## [17.3.1] - 2021-08-19

### Fixed

  - Fix regression due to missing Nexe binaries in new operating system environment. (17.3.0 did not contain the Nexe binaries and was reverted within minutes.)

## [17.3.0] - 2021-08-19 (reverted release)

### Updated

  - Version of Owncast installed via `site enable --owncast` upgraded to 0.0.8 (was 0.0.7).

## [17.2.1] - 2021-06-18

### Fixed

  - WebSockets are now being proxied correctly. Among other things, this fixes the issue with viewer counts appearing as zero (and chat not working) in Owncast installations.

## [17.2.0] - 2021-06-16

### Added

  - The new `--allow-embeds` option for the `serve` and `enable` commands allows site content to be embedded in other sites via iframes, etc. By default, this is off and an [X-Frame-Options](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options) header set to `SAMEORIGIN` is sent for security purposes to avoid click-jacking attacks, etc.

### Fixed

  - Owncast installations are run with the `--allow-embeds` option by default so live streams can now be embedded on other sites.

## [17.1.0] - 2021-06-16

### Updated

  - Now installs the latest [Owncast](https://owncast.online) ([version 0.0.7](https://github.com/owncast/owncast/releases/tag/v0.0.7)) when [setting up an Owncast server](https://owncast.online/quickstart/sitejs/) using `site enable --owncast`.

## [17.0.2] - 2021-04-29

### Fixed

  - Server now automatically restarts when middleware (in the _.middleware/_ folder of your project) is updated.

## [17.0.1] - 2021-04-27

### Fixed

  - Regression (in 17.0.0): Web Socket server is created only once when using both routes.js file (advanced routing) and DotJS together.

## [17.0.0] - 2021-04-25

### Breaking change


lib/Stats.js
////////////////////////////////////////////////////////////////////////////////
//
// Anonymous, ephemeral server statistics for Site.js
//
////////////////////////////////////////////////////////////////////////////////

const fs = require('fs')
const crypto = require('crypto')

const secretHexademicalStringOf32CharactersOrMore = /[0-9,a-z]{32,}/g

class Stats {

  constructor(statisticsRouteSettingFile = null) {
    this.startTime = new Date()

    // Sanity check.
    if (statisticsRouteSettingFile === null) {
      throw new Error('Error: Stats module constructor requires statistics route setting file path.')
    }

    this.requests = {}
    this.referrers = {}
    this.errors = {}
    this.responseCodes = {}
    this.missing = new Set()

    this.statisticsRouteSettingFile = statisticsRouteSettingFile
    this._route = null
  }


  // Force a new route to be generated.
  refreshRoute() {
    fs.unlinkSync(this.statisticsRouteSettingFile)
    return this.route
  }


  // The route for the view.
  get route () {
    if (this._route !== null) {
      return this._route
    }

    // We haven’t loaded the route yet. Load or create it.
    if (!fs.existsSync(this.statisticsRouteSettingFile)) {
      // Generate a random route.
      this._route = `/${crypto.randomBytes(16).toString('hex')}`
      fs.writeFileSync(this.statisticsRouteSettingFile, this._route)
    } else {
      // Load the existing route.
      this._route = fs.readFileSync(this.statisticsRouteSettingFile, 'utf-8')
    }

    return this._route
  }


  // Returns the middleware that captures the statistics.
  get middleware () {
    return (request, response, next) => {
      let referrer = request.headers.referer

      if (referrer !== undefined) {
        referrer = referrer.replace(secretHexademicalStringOf32CharactersOrMore, '…')
        this.referrers[referrer] === undefined ? this.referrers[referrer] = 1 : this.referrers[referrer]++
      }

      response.on('finish', () => {
        // Certain requests can contain sensitive information (like secret cryptographically-secure
        // random URLs). Make sure we hide these before saving the request path.
        let requestPath = request.path
          .replace(this.route, 'This page')
          .replace(secretHexademicalStringOf32CharactersOrMore, '…')


        if (response.statusCode === 404) {
          this.missing.add(requestPath)
        }
        this.requests[requestPath] === undefined ? this.requests[requestPath] = 1 : this.requests[requestPath]++
      })

      response.on('error', (error) => {
        console.log('Error, should we add this to stats?', error)
      })

      next()
    }
  }


  // Returns the view that renders the statistics.
  // TODO: Implement as WebSocket for live stats.
  get view () {
    return (request, response) => {

      const requestKeysToHtml = (keys) => {
        return keys.reduce((list, key) => {

          let route = key
          if (this.missing.has(route)) {
            // Highlight the missing routes in red.
            route = `<span style='color: red;'>${route}</span>`
          }

          return list += `<li>${route}: ${this.requests[key]}</li>`
        }, '')
      }

      const referrersToHtml = (keys) => {
        return keys.reduce((list, key) => list += `<li>${key}: ${this.referrers[key]}</li>`, '')
      }


      const totalRequests = Object.keys(this.requests).reduce((total, key) => total += this.requests[key], 0)
      const sortedRequestKeys = Object.keys(this.requests).sort((a,b) => -(this.requests[a] - this.requests[b]))
      const topThreeRequestKeys = sortedRequestKeys.slice(0, 3)

      const sortedReferrers = Object.keys(this.referrers).sort((a, b) => -(this.referrers[a] - this.referrers[b]))

      const administrationFragmentRegExp = /^\/admin.*?$/
      const none = '<li>None yet.</li>'
      const allRequestsList = requestKeysToHtml(sortedRequestKeys).replace(this.route, 'This page').replace(administrationFragmentRegExp, 'Administration section') || none
      const topThreeRequestsList = requestKeysToHtml(topThreeRequestKeys).replace(this.route, 'This page').replace(administrationFragmentRegExp, 'Administration section') || none

      const sortedReferrersList = referrersToHtml(sortedReferrers) || none

      const missingRequestsList = Array.from(this.missing).reduce((list, item) => list += `<li>${item}</li>`, '') || none

      response.status(200)
      .type('html')
      .end(`
        <!doctype html>
        <html lang="en">
        <head>
          <meta charset="utf-8">
          <meta name="viewport" content="width=device-width, initial-scale=1.0">
          <style>
            html {font-family: sans-serif; background-color: #eae7e1}
            body {padding-left: 2vw; padding-right: 2vw;}
            .requests { font-size: 10vw; font-weight: 200; line-height: 1; margin: 0; }
            ul { list-style: none; padding-left: 0; }
            li { padding: 0.25em 0; }
            li:nth-of-type(2n) { background-color: lightgray; }
            input[type='submit'] {
              font-size: 1.25em;
              background-color: #333;
              border: 1px solid black;
              border-radius: 0.25em;
              padding: 0.25em 2em;
              color: #eee;
            }
            code {
              font-size: 1.25em;
              background-color: #ccc;
              color: black;
              padding: 0.25em;
            }
          </style>
        </head>
        <body>
          <h1>Ephemeral Statistics</h1>

          <p>Powered by <a href='https://sitejs.org'>Site.js</a></p>

          <p><strong>Since:</strong> ${this.startTime}</p>

          <p><strong>To reset the statistics, restart the server using <code>site restart</code>.</strong></p>

          <p><a href='https://github.com/small-tech/site.js/blob/master/README.md#ephemeral-statistics'>Learn more about ephemeral statistics.</a></p>

          <form action ='${this.route}'><input type='submit' value='Load latest statistics' /></form>

          <h2 id='requests'>Requests</h2>

          <p class='requests'>${totalRequests}</p>

          <!-- List all requests, in order -->
          <ul>
            ${allRequestsList}
          </ul>

          <!-- 404s -->
          <h2 id='missing'>Missing</h2>
          <p>These are routes that have been requested but do not exist.</p>
          <ul>
            ${missingRequestsList}
          </ul>

          <h2 id='referrers'>Referrers</h2>
          <p>These are the URLs browsers have told us that links originate from.</p>
          <ul>
            ${sortedReferrersList}
          </ul>
        </body>

        <script>
          const secretHexademicalStringOf32CharactersOrMore = /[0-9,a-z]{32,}/g

          if (secretHexademicalStringOf32CharactersOrMore.exec(window.location) !== null) {
            // Hide the secret URL fragment so it is not accidentally displayed
            // in screenshots and is not added to the browser history.
            history.replaceState({}, '', window.location.href.replace(secretHexademicalStringOf32CharactersOrMore, 'stats'))
          }
        </script>
        </html>
      `)
    }
  }
}

module.exports = Stats
